Logging: 'test_update_sink' flakes with '503 Service Unavailable'

From: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/8002

```python
_________________________ TestLogging.test_update_sink _________________________

args = (parent: "projects/precise-truck-742"
sink {
  name: "test-update-sink-8002-1536695748"
  destination: "storage.googleapis.com/g-c-python-testing-8002-1536695748"
  filter: "logName:syslog AND severity>=INFO"
}
,)
kwargs = {'metadata': [('x-goog-api-client', 'gl-python/3.6.0 grpc/1.15.0rc1 gax/1.4.0 gapic/1.6.0 gccl/1.6.0')], 'timeout': 30.0}

    @six.wraps(callable_)
    def error_remapped_callable(*args, **kwargs):
        try:
>           return callable_(*args, **kwargs)

../api_core/google/api_core/grpc_helpers.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <grpc._channel._UnaryUnaryMultiCallable object at 0x7f6a4e6604e0>
request = parent: "projects/precise-truck-742"
sink {
  name: "test-update-sink-8002-1536695748"
  destination: "storage.googleapis.com/g-c-python-testing-8002-1536695748"
  filter: "logName:syslog AND severity>=INFO"
}

timeout = 30.0
metadata = [('x-goog-api-client', 'gl-python/3.6.0 grpc/1.15.0rc1 gax/1.4.0 gapic/1.6.0 gccl/1.6.0')]
credentials = None

    def __call__(self, request, timeout=None, metadata=None, credentials=None):
        state, call, = self._blocking(request, timeout, metadata, credentials)
>       return _end_unary_response_blocking(state, call, False, None)

../.nox/sys-3-6/lib/python3.6/site-packages/grpc/_channel.py:532: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

state = <grpc._channel._RPCState object at 0x7f6a4e511828>
call = <grpc._cython.cygrpc.SegregatedCall object at 0x7f6a4e49d248>
with_call = False, deadline = None

    def _end_unary_response_blocking(state, call, with_call, deadline):
        if state.code is grpc.StatusCode.OK:
            if with_call:
                rendezvous = _Rendezvous(state, call, None, deadline)
                return state.response, rendezvous
            else:
                return state.response
        else:
>           raise _Rendezvous(state, None, None, deadline)
E           grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
E           	status = StatusCode.UNAVAILABLE
E           	details = "The service is currently unavailable."
E           	debug_error_string = "{"created":"@1536695824.773475357","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1099,"grpc_message":"The service is currently unavailable.","grpc_status":14}"
E           >

../.nox/sys-3-6/lib/python3.6/site-packages/grpc/_channel.py:466: _Rendezvous

The above exception was the direct cause of the following exception:

self = <test_system.TestLogging testMethod=test_update_sink>

    def test_update_sink(self):
        SINK_NAME = 'test-update-sink%s' % (_RESOURCE_ID,)
        retry = RetryErrors(Conflict, max_tries=10)
        bucket_uri = self._init_storage_bucket()
        dataset_uri = self._init_bigquery_dataset()
        UPDATED_FILTER = 'logName:syslog'
        sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, bucket_uri)
        self.assertFalse(sink.exists())
>       retry(sink.create)()

tests/system/test_system.py:507: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../test_utils/test_utils/retry.py:95: in wrapped_function
    return to_wrap(*args, **kwargs)
google/cloud/logging/sink.py:136: in create
    unique_writer_identity=unique_writer_identity,
google/cloud/logging/_gapic.py:217: in sink_create
    unique_writer_identity=unique_writer_identity
google/cloud/logging_v2/gapic/config_service_v2_client.py:424: in create_sink
    request, retry=retry, timeout=timeout, metadata=metadata)
../api_core/google/api_core/gapic_v1/method.py:139: in __call__
    return wrapped_func(*args, **kwargs)
../api_core/google/api_core/retry.py:260: in retry_wrapped_func
    on_error=on_error,
../api_core/google/api_core/retry.py:177: in retry_target
    return target()
../api_core/google/api_core/timeout.py:206: in func_with_timeout
    return func(*args, **kwargs)
../api_core/google/api_core/grpc_helpers.py:61: in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value = None
from_value = <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "The service is currently unavai.../core/lib/surface/call.cc","file_line":1099,"grpc_message":"The service is currently unavailable.","grpc_status":14}"
>

>   ???
E   google.api_core.exceptions.ServiceUnavailable: 503 The service is currently unavailable.

<string>:3: ServiceUnavailable
```

--- a/logging/tests/system/test_system.py
+++ b/logging/tests/system/test_system.py
@@ -330,7 +330,10 @@ class TestLogging(unittest.TestCase):
         metric = Config.CLIENT.metric(
             METRIC_NAME, DEFAULT_FILTER, DEFAULT_DESCRIPTION)
         self.assertFalse(metric.exists())
-        metric.create()
+        retry = RetryErrors(Conflict)
+
+        retry(metric.create)()
+
         self.to_delete.append(metric)
         self.assertTrue(metric.exists())
 
@@ -342,10 +345,13 @@ class TestLogging(unittest.TestCase):
         before_metrics = list(Config.CLIENT.list_metrics())
         before_names = set(before.name for before in before_metrics)
         self.assertFalse(metric.name in before_names)
-        metric.create()
+        retry = RetryErrors(Conflict)
+        retry(metric.create)()
         self.to_delete.append(metric)
         self.assertTrue(metric.exists())
+
         after_metrics = list(Config.CLIENT.list_metrics())
+
         after_names = set(after.name for after in after_metrics)
         self.assertTrue(metric.name in after_names)
 
@@ -359,7 +365,9 @@ class TestLogging(unittest.TestCase):
         self.to_delete.append(metric)
         metric.filter_ = 'logName:other'
         metric.description = 'local changes'
+
         metric.reload()
+
         self.assertEqual(metric.filter_, DEFAULT_FILTER)
         self.assertEqual(metric.description, DEFAULT_DESCRIPTION)
 
@@ -375,7 +383,9 @@ class TestLogging(unittest.TestCase):
         self.to_delete.append(metric)
         metric.filter_ = NEW_FILTER
         metric.description = NEW_DESCRIPTION
+
         metric.update()
+
         after_metrics = list(Config.CLIENT.list_metrics())
         after_info = {metric.name: metric for metric in after_metrics}
         after = after_info[METRIC_NAME]
@@ -409,7 +419,9 @@ class TestLogging(unittest.TestCase):
         retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, uri)
         self.assertFalse(sink.exists())
+
         retry(sink.create)()
+
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
 
@@ -438,7 +450,9 @@ class TestLogging(unittest.TestCase):
         retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, TOPIC_URI)
         self.assertFalse(sink.exists())
+
         retry(sink.create)()
+
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
 
@@ -471,7 +485,9 @@ class TestLogging(unittest.TestCase):
         uri = self._init_bigquery_dataset()
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, uri)
         self.assertFalse(sink.exists())
+
         retry(sink.create)()
+
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
 
@@ -487,7 +503,9 @@ class TestLogging(unittest.TestCase):
         retry(sink.create)()
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
+
         after_sinks = list(Config.CLIENT.list_sinks())
+
         after_names = set(after.name for after in after_sinks)
         self.assertTrue(sink.name in after_names)
 
@@ -501,7 +519,9 @@ class TestLogging(unittest.TestCase):
         self.to_delete.append(sink)
         sink.filter_ = 'BOGUS FILTER'
         sink.destination = 'BOGUS DESTINATION'
+
         sink.reload()
+
         self.assertEqual(sink.filter_, DEFAULT_FILTER)
         self.assertEqual(sink.destination, uri)
 
@@ -517,7 +537,9 @@ class TestLogging(unittest.TestCase):
         self.to_delete.append(sink)
         sink.filter_ = UPDATED_FILTER
         sink.destination = dataset_uri
+
         sink.update()
+
         self.assertEqual(sink.filter_, UPDATED_FILTER)
         self.assertEqual(sink.destination, dataset_uri)
 

--- a/logging/tests/system/test_system.py
+++ b/logging/tests/system/test_system.py
@@ -341,7 +341,7 @@ class TestLogging(unittest.TestCase):
         self.assertFalse(metric.exists())
         before_metrics = list(Config.CLIENT.list_metrics())
         before_names = set(before.name for before in before_metrics)
-        self.failIf(metric.name in before_names)
+        self.assertFalse(metric.name in before_names)
         metric.create()
         self.to_delete.append(metric)
         self.assertTrue(metric.exists())
@@ -406,9 +406,10 @@ class TestLogging(unittest.TestCase):
         uri = self._init_storage_bucket()
         SINK_NAME = 'test-create-sink-bucket%s' % (_RESOURCE_ID,)
 
+        retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, uri)
         self.assertFalse(sink.exists())
-        sink.create()
+        retry(sink.create)()
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
 
@@ -434,9 +435,10 @@ class TestLogging(unittest.TestCase):
 
         TOPIC_URI = 'pubsub.googleapis.com/%s' % (topic_path,)
 
+        retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, TOPIC_URI)
         self.assertFalse(sink.exists())
-        sink.create()
+        retry(sink.create)()
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
 
@@ -465,22 +467,24 @@ class TestLogging(unittest.TestCase):
 
     def test_create_sink_bigquery_dataset(self):
         SINK_NAME = 'test-create-sink-dataset%s' % (_RESOURCE_ID,)
+        retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         uri = self._init_bigquery_dataset()
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, uri)
         self.assertFalse(sink.exists())
-        sink.create()
+        retry(sink.create)()
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
 
     def test_list_sinks(self):
         SINK_NAME = 'test-list-sinks%s' % (_RESOURCE_ID,)
         uri = self._init_storage_bucket()
+        retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, uri)
         self.assertFalse(sink.exists())
         before_sinks = list(Config.CLIENT.list_sinks())
         before_names = set(before.name for before in before_sinks)
-        self.failIf(sink.name in before_names)
-        sink.create()
+        self.assertFalse(sink.name in before_names)
+        retry(sink.create)()
         self.to_delete.append(sink)
         self.assertTrue(sink.exists())
         after_sinks = list(Config.CLIENT.list_sinks())
@@ -489,7 +493,7 @@ class TestLogging(unittest.TestCase):
 
     def test_reload_sink(self):
         SINK_NAME = 'test-reload-sink%s' % (_RESOURCE_ID,)
-        retry = RetryErrors(Conflict)
+        retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         uri = self._init_bigquery_dataset()
         sink = Config.CLIENT.sink(SINK_NAME, DEFAULT_FILTER, uri)
         self.assertFalse(sink.exists())
@@ -503,7 +507,7 @@ class TestLogging(unittest.TestCase):
 
     def test_update_sink(self):
         SINK_NAME = 'test-update-sink%s' % (_RESOURCE_ID,)
-        retry = RetryErrors(Conflict, max_tries=10)
+        retry = RetryErrors((Conflict, ServiceUnavailable), max_tries=10)
         bucket_uri = self._init_storage_bucket()
         dataset_uri = self._init_bigquery_dataset()
         UPDATED_FILTER = 'logName:syslog'

